src/main/java/org/truffleruby/interop/ForeignReadStringCachedHelperNode.java

```diff
@@ -99,20 +113,11 @@ public Object index(
             @Cached("create()") ForeignToRubyNode nameToRubyNode,
             @Cached("createBinaryProfile()") ConditionProfile arrayProfile,
             @Cached("createBinaryProfile()") ConditionProfile validArrayIndexProfile) {
-        if (arrayProfile.profile(RubyGuards.isRubyArray(receiver))) {
-            if (validArrayIndexProfile.profile(name instanceof Integer ||
-                    (name instanceof Long && RubyGuards.fitsInInteger((long) name)))) {
-                return getCallNode().call(frame, receiver, "[]", name);
-            } else {
-                throw UnknownIdentifierException.raise(toString(name));
-            }
-        } else {
-            return getCallNode().call(frame, receiver, "[]", nameToRubyNode.executeConvert(name));
-        }
+        return getCallNode().call(frame, receiver, INDEX_METHOD_NAME, nameToRubyNode.executeConvert(name));
```

Comment:
This looks identical to the specialization above. Does it matter if the `stringName` method exists if `INDEX_METHOD_NAME` exists?

Reply:
The guards are different. If you don't specify every combination of the two options (method defined and indexing method defined) you can pick one specialisation and then never go back to check if a new object actually fit the previous specialisation.

This pattern ensures that `[]` has priority over a method.

Reply:
I think we're saying the same thing but coming to different conclusions. The result of `methodDefined(frame, receiver, stringName, getDefinedNode())` doesn't matter if `methodDefined(frame, receiver, INDEX_METHOD_NAME, getIndexDefinedNode())` is `true`. If you find it clearer to have all 4 cases, that's fine. But, I think you really only have 3 cases and doing the `methodDefined(frame, receiver, stringName, getDefinedNode())` here is a bit misleading.

I could just be reading this wrong though. GitHub diffs aren't the easiest things to work with. I defer to you.

Reply:
The four specialisations are:

`method & !index`, `method & index`, `!method & index`, `!method & !index`

Are you suggesting that 116, `!method & index`, isn't needed? Or something else?

Reply:
Correct. Rather than a full truth table, I think the logic should look like:
```
if index
  call index
else
  if method
    call method
  else
    raise UnknownIdentifierException
end
```

Reply:
Do you think the current truth table we have has the same semantics as that nested logic? Or are you saying there's an error in the code in the PR and the nested logic is the correct one?

I'd rather keep the truth table if it's just a style thing - it's flat, you can see the conditions for each specialisation on their own, and with GVN it should end up as the same kind of code.

Reply:
I agree with Kevin here, it seems extra complexity.
I would move the `methodDefined(frame, receiver, "[]")` guards upper, since the decision is first based on `[]` really (it's irrelevant whether there is a method of the same name if `[]` is defined).
```ruby
def READ(obj, name)
  if name.start_with?('@')
    obj.instance_variable_get(name)
  elsif obj.is_a? Array
  elsif obj.is_a? Hash
  elsif obj.respond_to?(:[])
    obj[name]
  elsif obj.respond_to?(name)
    obj.method(name)
  else
    raise UnknownIdentifierException
  end
end
```
And BTW, Array and Hash don't need special treatment it seems.
Looking at the code I just wrote, I think it could be a nice way to document the semantics too :)

Reply:
The code works - we can make changes like this later.